lib/aggregate/paradigm-assignment.ts

```diff
@@ -0,0 +1,376 @@
+          ok = true;
+          break;
+        }
+      }
+      if (!ok) {
+        throw Error("tie breaker exists but was not applied");
+      }
+    } else {
+      console.log(`${matches.length} matches for ${head}`);
+      ok = false;
+    }
+  } else {
+    console.log(`${matches.length} matches for ${head}`);
+    ok = false;
+  }
+
+  return { analysis, paradigm: paradigm ?? undefined, ok };
+}
+
+/**
+ * If the FST analysis matches, return {analysis, paradigm}. Otherwise return null.
+ *
+ * For example, a +V+II analysis will match to an entry with pos = VII, but
+ * not to one with pos = VAI.
+ */
+export function matchAnalysis(
+  analysis: Analysis,
+  { head, pos }: { head: string; pos: unknown }
+): { analysis: Analysis; paradigm: string | null } | null {
+  if (!pos || !(typeof pos === "string")) {
+    return null;
+  }
+
+  const [_prefixTags, _lemma, suffixTags] = analysis;
+
+  if (pos.startsWith("I")) {
+    if (suffixTags.includes("+Ipc")) {
+      return { analysis, paradigm: null };
+    }
+  }
+
+  if (
+    pos === "PrA" &&
+    PERSONAL_PRONOUNS.has(head) &&
+    suffixTags.includes("+Pron") &&
+    suffixTags.includes("+Pers")
+  ) {
+    return { analysis, paradigm: "personal-pronouns" };
+  }
+
+  if (
+    pos === "PrA" &&
+    DEMONSTRATIVE_PRONOUNS.has(head) &&
+    suffixTags.includes("+Pron") &&
+    suffixTags.includes("+Dem") &&
+    suffixTags.includes("+A")
+  ) {
+    return { analysis, paradigm: "demonstrative-pronouns" };
+  }
+
+  if (
+    pos === "PrI" &&
+    DEMONSTRATIVE_PRONOUNS.has(head) &&
+    suffixTags.includes("+Pron") &&
+    suffixTags.includes("+Dem") &&
+    suffixTags.includes("+I")
+  ) {
+    return { analysis, paradigm: "demonstrative-pronouns" };
+  }
+
+  if ((pos === "PrA" || pos === "PrI") && suffixTags.includes("+Pron")) {
+    return { analysis, paradigm: null };
+  }
+
+  const specificWordClass = pos.split("-")[0];
+
+  for (let [paradigmName, paradigmSpecificWordClass, paradigmTags] of [
+    ["NA", "NA", ["+N", "+A"]],
+    ["NI", "NI", ["+N", "+I"]],
+    ["NDA", "NDA", ["+N", "+A", "+D"]],
+    ["NDI", "NDI", ["+N", "+I", "+D"]],
+    ["VTA", "VTA", ["+V", "+TA"]],
+    ["VTI", "VTI", ["+V", "+TI"]],
+    ["VAI", "VAI", ["+V", "+AI"]],
+    ["VII", "VII", ["+V", "+II"]],
+  ] as [string, string, string[]][]) {
+    if (
+      specificWordClass === paradigmSpecificWordClass &&
+      intersection(paradigmTags, suffixTags).length === paradigmTags.length
+    ) {
+      return { analysis, paradigm: paradigmName };
+    }
+  }
+  return null;
+}
+
+function doAssignment(entries: NdjsonEntry[]) {
+  const unusedPersonalPronouns = new Set(PERSONAL_PRONOUNS);
+  const unusedDemonstrativePronouns = new Set(DEMONSTRATIVE_PRONOUNS);
+
+  for (const entry of entries) {
+    const head = entry.head.sro!;
+    const pos = entry.dataSources?.CW?.pos?.toString();
+
+    const isProbablyMorpheme =
+      head.startsWith("-") || (head.endsWith("-") && pos !== "IPV");
```

Comment:
I _think_ this will miss any dependent nouns, since those are always written as `-stem-` but they do have paradigms. I don't see a good solution to this right now though.

Reply:
For dependent nouns, we decided to use the first person possessed singular form as the lemma, e.g. 1) _nôhkom_ for 'my grandmother', with the stem _-ohkom-_, when there is no _mi-/someone's_ form, or 2) _mitâs_ 'someones's pants', when there is a _mi-/someone's_ form. Thus, the lemma is always a stand-alone form, without any hyphens, expect of course if an entry is a bound morpheme, as is the case with preverbs (and prenouns), where the entry head(word) is marked with a final hyphen. As far as I understand, we shouldn't be using the stem to determine the status of the entry (as wordform, lemma, phrase, or morpheme).